### PR TITLE
GEODE-7852: Adding a SocketFactory configuration for client Pools

### PR DESCRIPTION
**Pull request by upthewaterspout**
Adding the ability to provide a SocketFactory to a client pool for use in
creating sockets. Adding an implementation of this SocketFactory that
configures the pool to use an SNI proxy.

Adding an automated test of the SniSocketFactory that uses docker compose to launch the locator and server in docker containers, and uses a real SNI proxy to ensure that clients connect to those locators and servers through the proxy.

### CODE REVIEW

File: geode-assembly/src/acceptanceTest/resources/org/apache/geode/client/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#
+version: '3'
+services:
+  geode:
+    container_name: 'geode'
+    image: 'geode:develop'
+    expose:
+      - '10334'
+      - '40404'
+#   entrypoint: gfsh
```
Comment (pivotal-jbarrett) [inline geode-assembly/src/acceptanceTest/resources/org/apache/geode/client/docker-compose.yml]: Delete commented code.

File: geode-core/src/distributedTest/java/org/apache/geode/cache/client/SocketFactoryDUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  private void verifyConnection(ClientCache client) {
+    // Verify connectivity with servers
+    Object functionResult =
+        FunctionService.onServers(client).execute(new TestFunction()).getResult();
+
+    assertThat(functionResult).isEqualTo(Arrays.asList("test"));
+  }
+
+
+  public static class TestFunction implements Function, DataSerializable {
```
Comment (pivotal-jbarrett) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/client/SocketFactoryDUnitTest.java]: Please don't leave raw type warnings. Use proper generic types or suppress warnings if generics can't work here. 

```java
public static class TestFunction implements Function<Void> ... {
...
public void execute<FunctionContext<Void> context);
}
```
```

File: geode-core/src/distributedTest/java/org/apache/geode/cache/client/SocketFactoryDUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+
+    // Verify the socket factory switched the hostname so we can connect
+    verifyConnection(client);
+  }
+
+  private void verifyConnection(ClientCache client) {
+    // Verify connectivity with servers
+    Object functionResult =
+        FunctionService.onServers(client).execute(new TestFunction()).getResult();
```
Comment (pivotal-jbarrett) [inline geode-core/src/distributedTest/java/org/apache/geode/cache/client/SocketFactoryDUnitTest.java]: Cleanup rawtype warnings.

Comment (upthewaterspout) [reply geode-core/src/distributedTest/java/org/apache/geode/cache/client/SocketFactoryDUnitTest.java]: Done.

**Pull request merged**